### ISSUE TITLE
array bound checking

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#278]
Similar functions on arrays and sequences have different behaviour as regards bound checking. For example, fn:head() returns an empty sequence if the input is empty, while array:head() throws an error.

Sometimes we want the error, sometimes we don't, but this should be orthogonal to whether we are using sequences or arrays.

Is there a way we can adapt the sequence functions to throw an error, or adapt the array functions so they don't?

Various ideas have been put forward, including:

- add extra optional parameters to functions to select the behaviour
- mirror the relevant functions (e.g. into a different namespace) to create an alternative version with different behaviour
- add options to the static context (array-bound-checking=yes|no, sequence-bound-checking=yes|no) to switch the behaviour (with appropriate mechanisms in XQuery and XSLT -- and perhaps XPath -- to set these options)
- do nothing, let users solve the problem for themselves by writing user-defined functions.

Note 1: the difference in behaviour affects operators as well as functions: contrast `$seq[0]` (which returns ()) with `$array(0)` (throws error). Making `$seq[0]` throw an error would mean we have to define it in a different way, since the formulation `$seq[position()=$N]` is intrinsically error-free.

Note 2: if `$array(0)` doesn't throw an error, it's not immediately obvious what it should do. Returning an empty sequence isn't ideal because the empty sequence is a valid entry in an array. For maps we have the same problem, which is why we have two functions `map:get()` and `map:contains()`.

**Comment by michaelhkay** [comment]
Because this affects operators as well as functions, I think my preferred solution would be to add options to the static context.

Along with this, XQuery/XPath probably needs an option to set static context options at a finer-grained level (XSLT does it with attributes having nested scope, such as `[xsl]:default-collation`). Something like

`with (default-collation = "xyzzy", array-bound-checking=false()) {$array?0}`

or perhaps use the under-utilised pragma syntax.

**Comment by dnovatchev** [comment]
> Note 2: if `$array(0)` doesn't throw an error, it's not immediately obvious what it should do. Returning an empty sequence isn't ideal because the empty sequence is a valid entry in an array. For maps we have the same problem, which is why we have two functions `map:get()` and `map:contains()`.

This means that if we want in any such case the array to return _something_ and not raise an error, this _something_ must not be anything that is currently in the XDM, because an array member can contain anything that is described at present in the XDM.

In Haskell they have the concept of [_**bottom**_ : **⊥**](https://en.wikipedia.org/wiki/Bottom_type) .

This probably is very similar to **_`xs:error`_**

We could deliberately choose another path by using  something like **_reserved keys_** in future ( to be postulated in the future) XDM constructs, that will be defined as not be manipulatable by user's code.  Although I am an atheist, this reminds me of something as the God, the Bible, or any other similar "sacred " thing, that the human imagination has already produced and been using for ages.

Having restricted and forbidden to access, or at least to manipulate, objects and modes isn't a new idea at all in software. Take the OS and user code, kernel and user state, Python and Javascript objects, ..., etc.

Thus, in designing future XDM concepts, it will not be bad to reserve some zone or range that is restricted for normal use.

**Comment by michaelhkay** [comment]
If we were to decide to have a static context option to disable array bound checking, then I think we would want to define it as follows.

In situations where currently error FOAY0001 is raised:

- `array:get(N)`, `$array(N)`, and `$array?N` instead return (). This is consistent with map:get() where we provide map:contains() for disambiguation of this result; we could consider providing array:contains(N) for consistency.
- `array:put()`, `array:replace()`, `array:remove()`, `array:insert-before()` return the input array unchanged
- `array:subarray()` follows the pattern of `fn:subsequence()`, where out-of-range endpoints for the selection are effectively adjusted to the start or end of the array as appropriate
- `array:head()` returns ()
- `array:tail()` returns []
- `array:index-where` ignores any supplied position that is out of range

**Comment by michaelhkay** [comment]
I'm not convinced that the converse, adding an option to enable sequence bound checking, is viable; I suspect that so much code would break that no-one would ever use it. But who knows unless we try? Also, people wouldn't consider setting it until they hit a bug that it would have caught if enabled. It's rather like asking people to use schema-awareness because it will catch more errors; people aren't prepared to take the up-front cost.

If we did do it, how would we define it?

- `A[N]`, where N is numeric, throws an error if N is not equal to an integer in the range 1 to count(A). Note that we can't usefully do any checking if they choose to write `A[position()=N]`
- functions such as `remove()`, `subsequence()`, `insert-before()`, `items-at()` throw an error if a position argument is out of range

**Comment by dnovatchev** [comment]
> If we were to decide to have a static context option to disable array bound checking, then I think we would want to define it as follows.
> 
> In situations where currently error FOAY0001 is raised:
> 
> * `array:get(N)`, `$array(N)`, and `$array?N` instead return (). This is consistent with map:get() where we provide map:contains() for disambiguation of this result; we could consider providing array:contains(N) for consistency.

This is obviously ambiguous with the case when the **_`N`_**<sub>th</sub> member of **_`$array`_** contains **`()`** (the empty sequence)! 

It is simply not telling us anything definite!

> * `array:put()`, `array:replace()`, `array:remove()` return the input array unchanged

This would be a debugging nightmare.  Many errors will go unnoticed!


> * `array:insert-before()` follows the pattern of `fn:insert-before()` - the new content is added at the beginning or end, as appropriate

Again, this will hide bugs in the code. We already have **_`fn:array:append()`_** to add to an empty array or to append an array

> * `array:subarray()` follows the pattern of `fn:subsequence()`, where out-of-range endpoints for the selection are effectively adjusted to the start or end of the array as appropriate

Again, this will hide bugs in the code. We are doing a disservice to the XPath programmer.

> * `array:head()` returns ()

Then how do we know whether the array is really empty or its first member contains `()` (the empty sequence) ?

This would make the currently useful and very foundational function **_`fn:head()`_** ambiguous. unreliable and useless.

> * `array:tail()` returns []

Then how do we know whether the array is really empty or its last member contains `[]`  (the empty array)?

This would make the currently useful and very foundational function **_`fn:tail()`_** ambiguous. unreliable and useless.

> * `array:index-where` ignores any supplied position that is out of range

Again, this will hide bugs in the code. We are doing a disservice to the XPath programmer.

**To summarize**:

The proposed rules bring ambiguity that makes the currently useful functions useless.

It is simply not logical to damage something useful that we already have in XPath 3.1.

**Comment by michaelhkay** [comment]
@dnovatchev I thought you were the one advocating for closer symmetry between sequence and array operations? That's all I'm proposing.

**Comment by ChristianGruen** [comment]
> It is simply not logical to damage something useful that we already have in XPath 3.1.

Maybe we got you wrong in the weekly meeting? What would you propose?

**Comment by dnovatchev** [comment]
> @dnovatchev I thought you were the one advocating for closer symmetry between sequence and array operations? That's all I'm proposing.

@michaelhkay I have proposed that there could be an option for the sequence functions (`head`, `tail`, etc) to produce error instead of the empty sequence. And at the meeting on 11/29 I explicitly stressed the fact that the reverse is ambiguous (the array functions (`array:head`, `array:tail`, etc) returning anything in case of empty array.

Here is **_[an excerpt from the meeting minutes](url)_**:

- DN: There’s no way to make an array behave like a sequence. An array can contain an empty sequence and that makes the cases indistinguishable.

**Comment by dnovatchev** [comment]
> > It is simply not logical to damage something useful that we already have in XPath 3.1.
> 
> Maybe we got you wrong in the weekly meeting? What would you propose?

@ChristianGruen Well, I explicitly stressed this in the last meeting, as documented in the minutes:

Here is **_[an excerpt from the meeting minutes](https://github.com/qt4cg/qtspecs/issues/url)_**:


- DN: There’s no way to make an array behave like a sequence. An array can contain an empty sequence and that makes the cases indistinguishable.

I also discussed in **_[my first comment to this issue](https://github.com/qt4cg/qtspecs/issues/278#issuecomment-1331497272)_** the impossibility of the array functions of unambiguously returning any kind of XDM item. There I referred to **_Haskell's concept of [bottom : ⊥](https://en.wikipedia.org/wiki/Bottom_type)_** , and mentioned that we need to have some special, reserved values (in "God mode") that are not to be manipulatable "in user mode"

**Comment by ChristianGruen** [comment]
@dnovatchev So you think that arrays should stay untouched, but that the access to sequences should (optionally) be handled more strictly?

I can understand the rationale behind that, but it would be a profound change to the language, even if it was made optional.

Personally, I would rather like to have a more tolerant array handling. It's pretty cumbersome to write recursive code with array:head/array:tail, which raises errors if an array is empty.

**Comment by dnovatchev** [comment]
> @dnovatchev So you think that arrays should stay untouched, but that the access to sequences should (optionally) be handled more strictly?

@ChristianGruen.

Exactly.

> 
> I can understand the rationale behind that, but it would be a profound change to the language, even if it was made optional.
> 


Adding optional stricter behavior cannot affect existing XPath 3.1 code-bases, thus this change will be invisible in all existing code.

> Personally, I would rather like to have a more tolerant array handling. It's pretty cumbersome to write recursive code with array:head/array:tail, which raises errors if an array is empty.

I understand this. The problem is that returning any type of XDM item can happen with a non-empty array (that contains this item). If we really want to "soften" the behavior of array functions, we could make them return in edge cases something that is not an XDM item-type (at present), that is a future type that is unknown until now and that can only be used to indicate error cases. This is why I mentioned Haskell's bottom.

Imagine we introduce "rock-bottom" or "pit" 😄 

As for writing  recursion being "pretty cumbersome", it is always better to be informed about an error as soon as possible, rather than potentially hiding its existence. Sentinel programming is actually eliminating the possibility for error, not hiding the fact that an error exists.

**Comment by ChristianGruen** [comment]
> it is always better to be informed about an error as soon as possible rather than potentially hiding its existence. 

This rather depends on how a function or expression is defined: If a function returns an empty sequence, null, undefined, etc. if no value exists, it’s not an error by definition. Think of e.g. `map.get` in Java (and possibly C#): A map can contain `null` as value, and if the function call returns `null`, you'll have no idea if it's the value or a non-existing value. The functions could have been defined to raise an error if a value does not exist, but this would always require an additional `map.contains` call whenever you are not sure about the contents of your map.

Or think of the XPath expression `/*[. = 'x']`: Would you prefer to get an error if no element with value `x` exists?

**Comment by dnovatchev** [comment]
> > it is always better to be informed about an error as soon as possible rather than potentially hiding its existence.
> 
> This rather depends on how a function or expression is defined: If a function returns an empty sequence, null, undefined, etc. if no value exists, it’s not an error by definition. Think of e.g. `map.get` in Java (and possibly C#): A map can contain `null` as value, and if the function call returns `null`, you'll have no idea if it's the value or a non-existing value. The functions could have been defined to raise an error if a value does not exist, but this would always require an additional `map.contains` call whenever you are not sure about the contents of your map.
> 
> Or think of the XPath expression `/*[. = 'x']`: Would you prefer to get an error if no element with value `x` exists?

C# dictionary class:

`someDict[nonExistentKey]`  **_[throws](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.item?view=net-7.0#exceptions)_** a  **_[KeyNotFoundException](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.keynotfoundexception?view=net-7.0)_**.

C# Queue class:

`someEmptyQueue.Dequeue()`  **_[throws InvalidOperationException](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.queue-1.dequeue?view=net-7.0#exceptions)_**

The same for the **_[peek()](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.queue-1.peek?view=net-7.0#exceptions)_** method.

There are a `TryDequeue()` and `TryPeek()` methods and these are essentially different methods **_where the success of the operation is a different result from the returned value result_**. Maybe we could design some similar `array:TryXXX()` functions.

**Comment by liamquin** [comment]
Another approach might be to add an extra argument, array:get($array, $n, $default-value), but that only works for functions, not for predicates. And we don't have sequence:get() today.

**Comment by line-o** [comment]
@liamquin `fn:items-at()` is proposed (see #249). That is `sequence:get()` right?

**Comment by ChristianGruen** [comment]
@dnovatchev Thanks for giving me some insight into C#.

Talking about XPath: Would you prefer to get an error for `//text()` if a document has no text nodes? What would be your favorite result for `<x>A</x>[. = 'B']`?

**Comment by rhdunn** [comment]
I was thinking of using https://www.w3.org/TR/xquery-31/#id-option-declaration to change the state of the XQuery behaviour w.r.t. error checking. An equivalent `<xsl:option name="qname" value="expression"/>` XSLT 4.0 element could be added -- that would increase the equivalence between XSLT and XQuery. (It would also be nice if there was a function you could call to check the value of an option declared in the static context, e.g. `fn:option-value($name as xs:QName) as item()*`.)

I'm strongly against the option of adding a parameter to the various sequence-based functions to control the exception/no exception behaviour.

Even with some way of controlling this functionality, you have a lot of compatibility issues and questions:
1. Does the static context of the script/stylesheet apply to the functions/variables of the imported modules?
2. What if a variable is declared using one of the functions before the static context for that behaviour is set?

And from a development point of view, what happens if a module/library a project is using switches from the non-throwing to throwing behaviour. -- That is going to break existing scripts/stylesheets that depend on the calling libraries, etc. returning an empty sequence.

**Comment by michaelhkay** [comment]
Conventions for managing things in the static context are well established. In XQuery you set it module-wide by a declaration in the prolog, in XSLT you set it with hierarchic scope using an attribute such as [xsl:]array-bound-checking. Just like default-collation etc, it applies only within its static scope, i.e. not to imported modules. So the semantics of code in library modules would not be affected.

**Comment by dnovatchev** [comment]
> @dnovatchev Thanks for giving me some insight into C#.
> 
> Talking about XPath: Would you prefer to get an error for `//text()` if a document has no text nodes? What would be your favorite result for `<x>A</x>[. = 'B']`?

@ChristianGruen When one is expecting a successful result in the form of a non-empty sequence of nodes, then just an empty sequence is convenient to signal unsuccessful result.

Unfortunately, an array member's value can be any XDM object, including an empty sequence or an empty array.

This is why the `TryXXX()` methods in .NET have to return a **_composite result_**, whose first part indicates success or failure and whose second part should be taken as the successful result **only** if the first part of the result is: _Success_

We could also specify standard `TryXXX()` functions.

Do note that one can test an array for not being empty (using one of the proposed functions [**_array:exists()_** and **_array:empty()_**](https://github.com/qt4cg/qtspecs/issues/229)), and this is the way to write primitive recursion (using **_`array:head()`_**,  **_`array:tail()`_** and **_`array;empty()`_** - to test for the base of the recursion)

**Comment by dnovatchev** [comment]
> Conventions for managing things in the static context are well established. In XQuery you set it module-wide by a declaration in the prolog, in XSLT you set it with hierarchic scope using an attribute such as [xsl:]array-bound-checking. Just like default-collation etc, it applies only within its static scope, i.e. not to imported modules. So the semantics of code in library modules would not be affected.

@michaelhkay and @rhdunn ,

I believe that they did it right in .NET by introducing and using the **_`TryXXX()`_** methods that return a **composite result**, whose first part alone indicates the success or failure of the call, and the 2nd part contains the result **only** in case the first part is Success.

Similarly in our  case where no suitable single value can be returned unambiguously to indicate an error, we could have such `TryXXX()` functions returning a composite result.

Revoking the error-raising behavior of existing array functions would result in **hiding bugs** and making them more difficult and not obvious to be aware of, thus I consider such relaxing to be a disservice to the XPath programmer.

**Comment by rhdunn** [comment]
The proposed syntax in https://github.com/qt4cg/qtspecs/issues/37 is to use:
```xquery
let ${exists, value} := map:try-get($map, "key")
return if ($exists) then $value else "missing"
```
which is (would be) equivalent to:
```xquery
let $ret := map:try-get($map, "key")
let $exists := $ret?exists
let $value := $ret?value
return if ($exists) then $value else "missing"
```

Note that the `${...}` syntax mirrors maps and the variables correspond to keys. There are symmetric `$(...)` and `$[...]` expressions for sequences and arrays, along with the ability to use different variable names for map keys, to specify the expected variable types, and to select the rest of the map/sequence/array.

Note that the let clause is not a part of the if conditional (like a scoped declaration in some languages, e.g. Go, C++20, etc.). This is because in XPath/XQuery it is part of a LetExpr/FLWORExpr.

**Comment by rhdunn** [comment]
On this specific proposal, I think having both a get with default (as an additional parameter to get) and a try-get function that returns a `record(exists as xs:boolean, value as item()*)` are useful, as there are different use cases for both -- e.g. looking up a key in a URI where you have a default value (get with default), or in a recursive function where you want to recurse if you have a value, and stop the recursion if you don't.

**Comment by dnovatchev** [comment]
Hope this is syntactically correct if `try-get()` returns a record:

```xq
( map:try-get($map, "key")[?success]?result, xs:error ) [1]
```

**Comment by michaelhkay** [comment]
`( map:try-get($map, "key")[?success]?result, xs:error ) [1]`

Neat idea. Slight drawbacks: (a) it relies on ?result being a singleton, (b) it should be `fn:error()`, and (c) I would expect most implementations to do the right thing with `fn:error()[1]` but in a formal sense it's questionable because it's theoretically legal to optimise it away (*).

Better is:

` map:try-get($map, "key")[?success]?result otherwise fn:error()

* XPath §2.3.4 "If... a processor is able to establish that there are only two possible outcomes of evaluating Q, namely the value V or an error, then the processor may deliver the result V without evaluating further items in the operand E". This rule is well-intentioned to legalise lazy evaluation, but it technically allows all sorts of undesirable outcomes, including returning () as the result of a call on `fn:error()[1]`

**Comment by dnovatchev** [comment]
> `( map:try-get($map, "key")[?success]?result, xs:error ) [1]`
> 
> Neat idea. Slight drawbacks: (a) it relies on ?result being a singleton, 

The first component of the return of `map:try-get()` is by definition an `xs:boolean`, thus a singleton.


>(b) it should be `fn:error()`, and (c) I would expect most implementations to do the right thing with `fn:error()[1]` but in a formal sense it's questionable because it's theoretically legal to optimise it away (*).
> 
> **Update**: I missed that there's a bigger drawback to both this and my suggested improvement
> 
> `map:try-get($map, "key")[?success]?result otherwise fn:error()`
> 
> Both call` fn:error()` if "key" is present in the map and maps to an empty sequence.

This is why the `?result` component has to be an array with a single member that is the result. Thus for an actual result that is the empty sequence, the result will be: `[()]` .

> 
> Looking at such examples makes me lean even more towards `get-with-default`. In practice there's almost always some value that is never going to appear in the map. Indeed if you really need to you can force it:
> 
> ```
> let $fallback := <fallback/>
> let $result := map:get-with-default($map, "key", $fallback)
> return $result[not(. is $fallback)]
> ```

Yes, default becomes usable if we have a "magic value" 😄 

I actually and seriously propose that we reserve a range of magic values for future use.

**Comment by ChristianGruen** [comment]
I would suggest introducing a function item as third parameter to the existing `array:get` and `map:get` functions:

```xquery
let $array := array { 1 to 10 }
let $pos := 11
let $fallback := function() { -1 }
return array:get($map, $key, $fallback)
```

```xquery
let $map := map { 1: 'one' }
let $key := 2
let $fallback := function($key) { error((), 'No value exists for key ' || $key) }
return map:get($map, $key, $fallback)
```

The function item will only be evaluated if the array or map yields no value. It may return a default value, raise an error, write logs, etc. The `$key` parameter of the function item could be optional.

**Comment by ChristianGruen** [comment]
`map:try-get` could indeed be enriched with a higher-order function, but I rather intended to propose `map:get#3` as a functional/declarative alternative and to avoid additional checks on the result.

The following two invocations would be equivalent:

```xquery
let $fallback := function() { ... }
return (
  map:try-get($map, 'key')[?success]?result otherwise $fallback(),
  map:get($map, 'key', $fallback)
)
```

@dnovatchev But maybe you have cases in your mind in which the record representation would be advantageous?

**Comment by michaelhkay** [comment]
I find `map:get($map, $key, $fallback := ->{})` the most elegant proposal so far. It's a simple extension to what we have today, it gives you a choice of returning a default value or raising an error when the key isn't present, and it allows a "computed default" as suggested by the use cases for issue 105 (but without any data model change). The "default value" can either be something the application can use directly (like zero or an empty string), or it can be an application-defined magic value to indicate absence (for example NaN or -1 or a special anyURI). I think that probably provides all the options we need. It extends naturally to array:get(), and we could also choose to do the same thing on `fn:items-at()`.

**Comment by rhdunn** [comment]
Having the ability to have a context-dependent default value controlled via a fallback function (instead of having a single default) makes it possible to process URI parameters, missing JSON properties, etc. in a more controlled way. There, you would want to have different default values or behaviours depending on which key(s) you were accessing.

One case where `map:with-default` could be useful (if the default was a function taking the key, so you could provide context-based defaults) is in conjunction with the proposed map destructuring syntax in #37 as there you don't currently have direct access to/control of the underlying `map:get` calls. -- It should be possible to extend the proposed syntax to support specifying the default value, in which case that would be an expression that would be converted into an inline function that is passed to the fallback function.

**Comment by ndw** [comment]
Resolved by #289 I believe.
